sceptre/connection_manager.py

```diff
@@ -183,6 +183,10 @@ def _determine_session_args(
         sceptre_role = (
             self.sceptre_role if sceptre_role == self.STACK_DEFAULT else sceptre_role
         )
+        # For historical reasons, if all three values are "None", that means we default to the
+        # Stack's configuration.
+        if (profile, region, sceptre_role) == (None, None, None):
+            profile, region, sceptre_role = self.profile, self.region, self.sceptre_role
```

Comment:
This is the actual fix right here. When I changed the ConnectionManager, I accidentally omitted this situation coverage.

Before, the logic was:
* If the profile, region, and role were all None, then...
    * if the stack_name was cached...
        * use the cached profile, region, and role by that cached stack name
     * otherwise, use THIS stack's profile, region, and role
* Otherwise, use the explicitly specified profile, region and role


After v4 was released, the logic was:
* If the stack_name was cached...
    * if the profile, region, and role were all None, then...
        * use the cached profile, region and role
     * otherwise, use the  specified profile, region, and role, using the cached value when no value has been specified
         * _This was a gap in our old logic, so this change was intentional and fixed some things._
* otherwise, use the  specified profile, region, and role, using the current stack's value when no value has been specified

**The gap that I found that was causing some bugs** was when the stack_name was **not** cached, but the profile, region, and role were passed as None. It turns out the !stack_output_external resolver would trigger this exact situation. But this wouldn't actually show up in terms of blow-ups when the user had a profile with a region or a default region specified as an environment variable. 

With this fix, the logic is now:
* If the stack_name was cached...
    * if the profile, region, and role were all None, then...
        * use the cached profile, region and role
     * otherwise, use the  specified profile, region, and role, using the cached value when no value has been specified
         * _This was a gap in our old logic, so this change was intentional and fixed some things._
* otherwise...
    * If the profile, region, and role are all None, then...
        * Use the current_stack's profile, region, and role  
    * Otherwise, use the specified profile, region, and role, using the current stack's value when no value has been specified

Reply:
Great write-up!